# TFLearn
* TFlearn is a modular and transparent deep learning library built on top of Tensorflow. It was designed to provide a higher-level API to TensorFlow in order to facilitate and speed-up experimentations, while remaining fully transparent and compatible with it.

* I hope to get better results with less code than with pure TF 

### Import Tensorflow and TFLearn

In [2]:
import tflearn
from tflearn.layers.conv import conv_2d, max_pool_2d
from tflearn.layers.core import input_data, dropout, fully_connected
from tflearn.layers.estimator import regression
import tflearn.datasets.mnist as mnist

### Get data and reshape it for square since we know how it is stored

In [3]:
X, Y, test_x, test_y = mnist.load_data(one_hot=True)
X = X.reshape([-1, 28, 28, 1])
test_x = test_x.reshape([-1, 28, 28, 1])

Extracting mnist/train-images-idx3-ubyte.gz
Extracting mnist/train-labels-idx1-ubyte.gz
Extracting mnist/t10k-images-idx3-ubyte.gz
Extracting mnist/t10k-labels-idx1-ubyte.gz


### Define our network

In [4]:
net = input_data(shape=[None, 28, 28, 1], name='input')

net = conv_2d(net, 32, 2, activation='relu')
net = max_pool_2d(net, 2)

net = conv_2d(net, 64, 2, activation='relu')
net = max_pool_2d(net, 2)

net = conv_2d(net, 64, 2, activation='relu')
net = max_pool_2d(net, 2)

net = fully_connected(net, 1024, activation='relu')
net = dropout(net, 0.8)

net = fully_connected(net, 10, activation='softmax')
net = regression(net, optimizer='adam', learning_rate=0.001, loss='categorical_crossentropy', name='targets')

### Define model

In [5]:
model = tflearn.DNN(net)

### Train model

In [6]:
model.fit({'input': X}, {'targets': Y}, n_epoch=5, validation_set=({'input': test_x}, {'targets': test_y}), 
    snapshot_step=500, show_metric=True, run_id='MNIST')

Training Step: 4299  | total loss: 0.13431 | time: 46.403s
| Adam | epoch: 005 | loss: 0.13431 - acc: 0.9881 -- iter: 54976/55000
Training Step: 4300  | total loss: 0.12153 | time: 48.749s
| Adam | epoch: 005 | loss: 0.12153 - acc: 0.9893 | val_loss: 0.03462 - val_acc: 0.9898 -- iter: 55000/55000
--


### We have 98.93% with only 5 epochs 
Notice: Epoch took significantly more time than before